### PR TITLE
[XCMP Foreign Fee]: Foreign fee spike

### PR DESCRIPTION
**Pull request by atomaka**
* Apply patch from https://gist.github.com/atomaka/d4e63de2f3d743314ca2f79379a6a994
* Build
* Add TUR: https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9946#/extrinsics/decode/0x32010e000a000000184e61746976650c545552000000000000000000000000000000000101000001009e588e61000000000000000000000000000000000000000000
* Add UNIT: https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9946#/extrinsics/decode/0x32010e000a00000010556e697410554e49540000000000000000000000000000000001010001003d1f01009e588e61000000000000000000000000000000000000000000
* Allow UNIT off chain: https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9946#/extrinsics/decode/0x32013e00cf0700000200000000009e588e61000000000000000000000000ca9a3b000000000000000000000000
* Give Alice UNIT: https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9946#/extrinsics/decode/0x32010c0200d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d0200000000008a5d7845630100000000000000000000000000000000
* Super fake scheduled xcmp task: https://polkadot.js.org/apps/?rpc=ws%3A%2F%2F127.0.0.1%3A9946#/extrinsics/decode/0x3c020410040000000000000000cf0700000200000004100046c32300000000

TODO:

* Rename `DuplicateCurrencyAdapter`
* Actually convert the fees xD
Screen Shot 2022-09-22 at 20 17 32

### CODE REVIEW

File: runtime/neumann/src/fee_handler.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
+use crate::*;
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:1]: TODO

File: runtime/neumann/src/fee_handler.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+			)
+			.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::withdraw(currency_id, who, fee.into())
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::deposit(currency_id, &TA::get(), fee.into()) // treasury account
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			// TODO: Fire event for deposit
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:124]: TODO

File: runtime/neumann/src/fee_handler.rs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+			} else {
+				WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+			};
+
+			match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+				Ok(imbalance) => Ok(Some(imbalance)),
+				Err(_) => Err(InvalidTransaction::Payment.into()),
+			}
+		} else {
+			let currency_id = call_information.token_id.into();
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:108]: Then convert the currency using data in `call_informaton`

File: runtime/neumann/src/fee_handler.rs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+			let withdraw_reason = if tip.is_zero() {
+				WithdrawReasons::TRANSACTION_PAYMENT
+			} else {
+				WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+			};
+
+			match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+				Ok(imbalance) => Ok(Some(imbalance)),
+				Err(_) => Err(InvalidTransaction::Payment.into()),
+			}
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:106]: Code should match what occurs in  the `pallet_transaction_payment` `CurrencyAdapter`.

File: runtime/neumann/src/fee_handler.rs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+				who,
+				fee.saturating_add(MC::minimum_balance(currency_id).into()).into(),
+			)
+			.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::withdraw(currency_id, who, fee.into())
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::deposit(currency_id, &TA::get(), fee.into()) // treasury account
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:122]: Just handle `OnUnbalanced` for foreign tokens ourselves.

File: runtime/neumann/src/fee_handler.rs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+			// TODO: Fire event for deposit
+
+			// We dealt with imbalance here so don't let `correct_and_deposit_fee` do it
+			Ok(None)
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:127]: Prevent `correct_and_deposit_fee` from doing anything when tokens are foreign.

File: runtime/neumann/src/fee_handler.rs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+			// merge the imbalance caused by paying the fees and refunding parts of it again.
+			let adjusted_paid = paid
+				.offset(refund_imbalance)
+				.same()
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+			// Call someone else to handle the imbalance (fee and tip separately)
+			let (tip, fee) = adjusted_paid.split(tip);
+			OU::on_unbalanceds(Some(fee).into_iter().chain(Some(tip)));
+		}
+		Ok(())
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:161]: Matches `pallet_transaction_payment` `CurrencyAdapter`.

File: runtime/turing/src/lib.rs [hunk 8]
```diff
@@ -576,10 +576,147 @@ where
 	}
```
Comment (atomaka) [inline runtime/turing/src/lib.rs:576]: Ignore turing for now. Neumann will be copied in later.

File: runtime/neumann/src/fee_handler.rs [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+		} else {
+			FeeInformation::default()
+		}
+	}
+}
+pub type CallOf<T> = <T as frame_system::Config>::Call;
+type NegativeImbalanceOf<C, T> =
+	<C as Currency<<T as frame_system::Config>::AccountId>>::NegativeImbalance;
+
+pub struct DuplicateCurrencyAdapter<MC, C, OU, TA, FCP>(PhantomData<(MC, C, OU, TA, FCP)>);
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:52]: Name help xD

File: primitives/src/assets.rs [hunk 10]
```diff
@@ -58,3 +58,11 @@ pub struct CustomMetadata {
 	/// The token conversion rate of Native to Foreign, ie. 1::10
 	pub conversion_rate: Option<ConversionRate>,
 }
+impl CustomMetadata {
+	pub fn convert_from_foreign_to_native(&self, fee: u128) -> Option<u128> {
+		match self.conversion_rate {
+			Some(value) => Some(fee * value.native as u128 / value.foreign as u128),
```
Comment (atomaka) [inline primitives/src/assets.rs:64]: Almost certainly not correct.  How is `ConversonRate` intended to be used @lreesby ?

File: runtime/neumann/src/fee_handler.rs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+				foreign_fee
+					.saturating_add(MC::minimum_balance(currency_id).saturated_into())
+					.saturated_into(),
+			)
+			.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::withdraw(currency_id, who, foreign_fee.saturated_into())
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::deposit(currency_id, &TA::get(), foreign_fee.saturated_into()) // treasury account
```
Comment (atomaka) [inline runtime/neumann/src/fee_handler.rs:129]: @irsal Do we want foreign funds to go to a different treasury accont.

Comment (irsal) [reply runtime/neumann/src/fee_handler.rs:129]: what was the existing one? or are we talking about the actual treasury?

File: runtime/neumann/src/fee_handler.rs (hunk 2 above)
Comment (atomaka) [reply runtime/neumann/src/fee_handler.rs:132]:
```suggestion
```

File: runtime/neumann/src/fee_handler.rs [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+		}
+		Ok(())
+	}
+}
+
+pub trait FeeConversion {
+	fn convert_fee_into_foreign(&self, fee: Balance) -> Option<Balance>;
+}
+impl FeeConversion for orml_asset_registry::AssetMetadata<Balance, CustomMetadata> {
+	fn convert_fee_into_foreign(&self, fee: Balance) -> Option<Balance> {
```
Comment (nvengal) [inline runtime/neumann/src/fee_handler.rs:205]: Got you a failing test-
```
	#[test]
	fn what_happens_with_large_decimal_diff() {
		let asset = test_asset!(20, 10000:1);
		assert_eq!(asset.convert_fee_into_foreign(1000000000), Some(1000))
	}
```
integer math gets you in trouble when the decimal diff is large. I'm not sure how likely this is but I don't think the example I picked is totally crazy: token with 20 decimals that has a 10000:1 exchange rate with us.

File: runtime/neumann/src/fee_handler.rs (hunk 7 above)
Comment (nvengal) [reply runtime/neumann/src/fee_handler.rs:197]: Does it make sense to forward to the transaction payment pallet here rather than copying the code?
```
pallet_transaction_payment::CurrencyAdapter<Balances, DealWithInclusionFees<Runtime>>::correct_and_deposit_fee(...)
```

File: runtime/neumann/src/fee_handler.rs [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+			return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
+		}
+
+		let currency_id = call_information.token_id.into();
+		let foreign_fee: MC::Balance = call_information
+			.asset_metadata
+			.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?
+			.convert_fee_into_foreign(fee.saturated_into())
+			.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?
+			.saturated_into();
```
Comment (nvengal) [inline runtime/neumann/src/fee_handler.rs:93]: General question- does this code need to guard against the fee being `0`? If the fee is `0` does that imply that there was a math bug or misconfiguration? Basically, does a fee of `0` here present a security risk to our chain and should we explicitly deny it?

File: primitives/Cargo.toml [hunk 14]
```diff
@@ -20,6 +20,7 @@ default = ["std"]
 std = [
 	"codec/std",
 	"scale-info/std",
+	"sp-consensus-aura/std",
```
Comment (rhuttman) [inline primitives/Cargo.toml:23]: why did you need to add this? You aren't using it anywhere

File: runtime/neumann/src/fee_handler.rs [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+	MC: MultiCurrency<<T as frame_system::Config>::AccountId>,
+	MC::CurrencyId: From<TokenId>,
+	TA: Get<T::AccountId>,
+{
+	fn withdraw_fee(
+		who: &T::AccountId,
+		fee: MC::Balance,
+		call_information: FeeInformation,
+	) -> Result<(), TransactionValidityError> {
+		if call_information.xcm_data.is_none() || call_information.asset_metadata.is_none() {
```
Comment (rhuttman) [inline runtime/neumann/src/fee_handler.rs:83]: this doesn't seem needed as you also check for `None` on line 88

File: runtime/neumann/src/fee_handler.rs [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+		// orml_tokens doesn't provide withdraw that allows setting existence so prevent
+		// withdraw if they don't have enough to avoid reaping
+		MC::ensure_can_withdraw(
+			currency_id,
+			who,
+			foreign_fee.saturating_add(MC::minimum_balance(currency_id)),
+		)
+		.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+		MC::withdraw(currency_id, who, foreign_fee)
```
Comment (rhuttman) [inline runtime/neumann/src/fee_handler.rs:104]: remind me why we aren't just using a transfer?

File: runtime/neumann/src/fee_handler.rs (hunk 6 above)
Comment (rhuttman) [reply runtime/neumann/src/fee_handler.rs:163]: aren't you already doing that at the end of `ForeignFeeProcessor.withdraw_fee`?

File: runtime/neumann/src/fee_handler.rs [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+impl FeeConversion for orml_asset_registry::AssetMetadata<Balance, CustomMetadata> {
+	fn convert_fee_into_foreign(&self, fee: Balance) -> Option<Balance> {
+		let decimaled_fee = if self.decimals >= TOKEN_DECIMALS {
+			fee / 10_u128.pow(self.decimals - TOKEN_DECIMALS)
+		} else {
+			fee * 10_u128.pow(TOKEN_DECIMALS - self.decimals)
+		};
+
+		match self.additional.conversion_rate {
+			Some(value) => Some(decimaled_fee * value.native as Balance / value.foreign as Balance),
```
Comment (rhuttman) [inline runtime/neumann/src/fee_handler.rs:213]: we decided not to up-charge for this then?

File: primitives/Cargo.toml (hunk 14 above)
Comment (atomaka) [reply primitives/Cargo.toml:23]: I originally added the decimal conversion to the `CustomMetadata` struct. We were missing this dependency to be able to run the `primitives` module in isolation. To verify:

```
git checkout master; cd primitives; cargo check
```

File: runtime/neumann/src/fee_handler.rs (hunk 16 above)
Comment (atomaka) [reply runtime/neumann/src/fee_handler.rs:88]: `transfer` still doesn't allow setting existence.  I think we decided they were the same and it didn't matter.  I'm happy with whichever unless @nvengal remembers something we don't.

File: runtime/neumann/src/fee_handler.rs (hunk 6 above)
Comment (atomaka) [reply runtime/neumann/src/fee_handler.rs:145]: `ForeignFeeProcessor.withdraw_fee` returns an empty tuple and not an option.  We could change that, although I think I prefer the explicitness of this given how important returning `None` is.

File: runtime/neumann/src/fee_handler.rs (hunk 16 above)
Comment (nvengal) [reply runtime/neumann/src/fee_handler.rs:88]: My opinion, loosely held, is that it's still best to stick with the pattern of `withdraw` then `deposit` for handling fees. The pattern for fees seems to be withdraw and then handle or burn imbalance. I think there's value in sticking to the general pattern. But, again, loosely held. I'm having trouble figuring out if `transfer` has any read/write savings associated.

**Comment by nvengal** [comment]
Moving to draft for now. Open questions need to be resolved once xcmp foreign fee work is resumed.

**Comment by nvengal** [comment]
outdated
